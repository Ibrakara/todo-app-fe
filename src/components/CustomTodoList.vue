<template>
  <div class="todo-list">
    <div
      v-for="todo in todos"
      :key="todo._id"
      class="todo-item"
      :class="{ completed: todo.completed }"
    >
      <span
        :class="{ completed: todo.completed }"
        class="todo-text"
        @click="$emit('toggleComplete', todo._id)"
      >
        {{ todo.title }}
      </span>
      <button @click="$emit('deleteTodo', todo._id)" class="delete-button">
        Delete
      </button>
    </div>
  </div>
</template>

<script>
export default {
  name: "CustomTodoList",
  props: {
    todos: {
      type: Array,
      default: () => [],
    },
  },
};
</script>

<style scoped>
.todo-list {
  margin-top: 20px;
  max-width: 500px;
  margin-left: auto;
  margin-right: auto;
}
.todo-item {
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #1c1f2b;
  padding: 10px 20px;
  margin-bottom: 10px;
  border-radius: 10px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
  color: #fefefe;
  transition: background-color 200ms ease;
}
.todo-item.completed {
  background-color: #1dd1a1;
  color: #ffffffcc;
}
.todo-text {
  flex-grow: 1;
  text-align: left;
  color: #fefefe;
  cursor: pointer;
}
.todo-text.completed {
  text-decoration: line-through;
}
.delete-button {
  background-color: #ff6b6b;
  border: none;
  border-radius: 5px;
  color: white;
  padding: 5px 10px;
  cursor: pointer;
  transition: background-color 200ms ease, transform 150ms ease;
}
.delete-button:hover {
  background-color: #ff4c4c;
  transform: scale(1.1);
}
.delete-button:active {
  transform: scale(0.95);
}
</style>
