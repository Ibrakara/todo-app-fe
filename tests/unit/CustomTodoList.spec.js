import { mount } from "@vue/test-utils";
import CustomTodoList from "../../src/components/CustomTodoList.vue";

describe("CustomTodoList", () => {
  const mockTodos = [
    { _id: "1", title: "First todo", completed: false },
    { _id: "2", title: "Second todo", completed: true },
  ];

  it("renders correctly with todos", () => {
    const wrapper = mount(CustomTodoList, {
      propsData: {
        todos: mockTodos,
      },
    });

    const todoItems = wrapper.findAll(".todo-item");
    expect(todoItems.length).toBe(mockTodos.length);

    expect(todoItems.at(0).text()).toContain("First todo");
    expect(todoItems.at(1).text()).toContain("Second todo");
  });

  it("renders correctly with empty todos", () => {
    const wrapper = mount(CustomTodoList, {
      propsData: {
        todos: [],
      },
    });
    expect(wrapper.findAll(".todo-item").length).toBe(0);
  });

  it("applies completed class for completed todos", () => {
    const wrapper = mount(CustomTodoList, {
      propsData: {
        todos: mockTodos,
      },
    });

    const todoItems = wrapper.findAll(".todo-item");
    expect(todoItems.at(0).classes()).not.toContain("completed");
    expect(todoItems.at(1).classes()).toContain("completed");
  });

  it("emits toggleComplete event when todo text is clicked", async () => {
    const wrapper = mount(CustomTodoList, {
      propsData: {
        todos: mockTodos,
      },
    });

    await wrapper.findAll(".todo-text").at(0).trigger("click");
    expect(wrapper.emitted().toggleComplete).toBeTruthy();
    expect(wrapper.emitted().toggleComplete[0]).toEqual(["1"]);
  });

  it("emits deleteTodo event when delete button is clicked", async () => {
    const wrapper = mount(CustomTodoList, {
      propsData: {
        todos: mockTodos,
      },
    });

    await wrapper.findAll(".delete-button").at(0).trigger("click");
    expect(wrapper.emitted().deleteTodo).toBeTruthy();
    expect(wrapper.emitted().deleteTodo[0]).toEqual(["1"]);
  });

  it("applies correct styling classes", () => {
    const wrapper = mount(CustomTodoList, {
      propsData: {
        todos: mockTodos,
      },
    });

    const completedTodo = wrapper.findAll(".todo-item").at(1);
    expect(completedTodo.find(".todo-text").classes()).toContain("completed");
    expect(completedTodo.classes()).toContain("completed");

    expect(wrapper.findAll(".delete-button").length).toBe(mockTodos.length);
  });

  it("matches snapshot", () => {
    const wrapper = mount(CustomTodoList, {
      propsData: {
        todos: mockTodos,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
