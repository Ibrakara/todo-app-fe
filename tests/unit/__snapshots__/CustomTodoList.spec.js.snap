// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTodoList matches snapshot 1`] = `
<div class="todo-list">
  <div class="todo-item"><span class="todo-text">
      First todo
    </span> <button class="delete-button">
      Delete
    </button></div>
  <div class="todo-item completed"><span class="todo-text completed">
      Second todo
    </span> <button class="delete-button">
      Delete
    </button></div>
</div>
`;
